src/security/CertGadgets.cc

```diff
@@ -35,21 +35,21 @@ Security::CertIssuerName(const CertPointer &cert)
     gnutls_x509_dn_t dn;
     auto x = gnutls_x509_crt_get_issuer(cert.get(), &dn);
     if (x != GNUTLS_E_SUCCESS) {
-        debugs(83, DBG_PARSE_NOTE(2), "WARNING: issuer: " << Security::ErrorString(x));
+        debugs(83, DBG_PARSE_NOTE(2), "WARNING: cannot get certificate Issuer: " << Security::ErrorString(x));
         return out;
     }
 
     gnutls_datum_t str;
     x = gnutls_x509_dn_get_str(dn, &str);
     if (x != GNUTLS_E_SUCCESS) {
-        debugs(83, DBG_PARSE_NOTE(2), "WARNING: issuer: " << Security::ErrorString(x));
+        debugs(83, DBG_PARSE_NOTE(2), "WARNING: cannot get certificate Issuer: " << Security::ErrorString(x));
         return out;
     }
     out.append(reinterpret_cast<const char *>(str.data), str.size);
     gnutls_free(str.data);
 
 #else
-    out.append("[not implemented]");
+    debugs(83, DBG_PARSE_NOTE(2), "WARNING: cannot get certificate Issuer: [not implemented]");
```

Comment:
Can a properly functioning Squid ever reach this (and similar) code? We have a bunch of functions that treat this no-library clause  differently: some do nothing at all, some log a problem, some return an error but log nothing, etc. I wonder whether there is a single good handling policy applicable to all of them that we can start using going forward?

For example, what do you think about throwing in virtually all those no-library places (in new/changed code) because a properly functioning Squid will never call these low-level TLS manipulation functions without being built with a TLS library?

P.S. If we can find such a policy, we can wrap it in a trivial reusable/documented function to be called here and elsewhere, of course.

Reply:
Ideally / long-term throwing would be right. But currently these calls are made by a mix of callers from cert mimic to ACLs with different abilities to handle (or not) the exceptions. We need the wider Squid code to be much cleaner before this can reliably be done.

For now, this seems to me the best implementation for these two particular. Empty string is a valid output from here (meaning "unable to find the data"). Logging that the reason it could not be found was library API not implemented.

Other API functions/methods have some variance on whether that is valid behaviour. So will need re-evaluation under the new principles to see whether they need changing, and how. That is out of scope for now.

Reply:
> Alex: Can a properly functioning Squid ever reach this (and similar) code?

> Amos: Ideally / long-term throwing would be right. But currently these calls are made by a mix of callers from cert mimic to ACLs with different abilities to handle (or not) the exceptions. We need the wider Squid code to be much cleaner before this can reliably be done.

I could not find the answer to my first question in your response, so I will try to enumerate all possible answers. Assuming correct callers (and we control them):

* If this `#else` code should never be reached, then why provide an API for the no-library cases that do not need it (and that get exposed to dangers if they accidentally access it)?
* If this `#else` code should be reachable by any caller, then why are we warning about such cases?
* If this `#else` reachability depends on the caller, then we seem to be creating a dangerous API that, depending on the caller, can result in (using your example) a silently mismatching ACL.

How about hiding this certificate interrogation API from non-library code? AFAICT, it is the _safest_ option as far as runtime is concerned. Would it create serious problems for any existing callers? If it will, perhaps it would be easier to solve _those_ (more specific/isolated) problems instead?